backend/persistent/plotly-api.js

```diff
@@ -0,0 +1,254 @@
+    if (apiKey) {
+        authorization = 'Basic ' + new Buffer(
+            username + ':' + apiKey
+        ).toString('base64');
+    } else if (accessToken) {
+        authorization = `Bearer ${accessToken}`;
+    } else {
+        throw new Error('Missing apiKey or accessToken');
+    }
+    return fetch(`${getSetting('PLOTLY_API_URL')}/v2/${relativeUrl}`, {
+        method,
+        headers: {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json',
+            'Plotly-Client-Platform': 'db-connect',
+            'Authorization': authorization
+        },
+        body: body ? JSON.stringify(body) : null
+    });
+}
+
+export function newDatacache(payloadJSON, type, requestor) {
+    const {username, apiKey, accessToken} = getCredentials(requestor);
+
+    const body = new FormData();
+    body.append('type', type);
+    body.append('origin', 'Falcon');
+    body.append('payload', payloadJSON);
+
+    if (username) {
+        body.append('username', username);
+    }
+
+    /*
+     * Authentication is only required for on-premise private-mode for this
+     * endpoint, so even if the user is not logged in, we should still be able
+     * to proceed with blank `Authorization` header.
+     */
+    let authorization = '';
+    if (apiKey) {
+        authorization = 'Basic ' + new Buffer(
+            requestor + ':' + apiKey
+        ).toString('base64');
+    } else if (accessToken) {
+        authorization = `Bearer ${accessToken}`;
+    }
+
+    return fetch(`${getSetting('PLOTLY_URL')}/datacache`, {
+        method: 'POST',
+        body: body,
+        headers: {
+            'Authorization': authorization
+        }
+    }).then(res => {
+        if (res.status !== 200) {
+            return res.text().then(text => {
+                throw new Error(`Failed request 'datacache'. Status: ${res.status}. Body: ${text}`);
+            });
+        }
+        return res.json();
+    }).catch(err => {
+        Logger.log(err, 0);
+        throw err;
+    });
+}
+
+export function getCurrentUser(requestor) {
+    const {username, apiKey, accessToken} = getCredentials(requestor);
+
+    return plotlyAPIRequest('users/current', {
+        method: 'GET',
+        username,
+        apiKey,
+        accessToken
+    });
+}
+
+export function newGrid(filename, columnnames, rows, requestor) {
+    const {username, apiKey, accessToken} = getCredentials(requestor);
+
+    const columns = getColumns(rows, columnnames.length);
+
+    const cols = {};
+    columnnames.forEach((name, i) => {
+        cols[name] = {'data': columns[i], order: i};
+    });
+    const grid = {cols};
+
+    return plotlyAPIRequest('grids', {
+        method: 'POST',
+        username,
+        apiKey,
+        accessToken,
+        body: {
+            data: grid,
+            world_readable: true,
+            parent: -1,
+            filename
+        }
+    });
+}
+
+export function getGridMetadata(fid, requestor) {
```

Comment:
Should we use fid or filename?  Above it is filename and now it is fid

Reply:
Plotly's API uses `fid`.

Reply:
we can use `getGridMeta` as it is more consistent with Plotly nomenclature. Not a big deal though..